crates/ruff/src/rules/refurb/rules/repeated_append.rs

```diff
@@ -0,0 +1,377 @@
+    }
+
+    fn autofix_title(&self) -> Option<String> {
+        let RepeatedAppend { replacement, .. } = self;
+        Some(format!("Replace with `{replacement}`"))
+    }
+}
+
+pub(crate) fn convert_consecutive_appends_into_extend(checker: &mut Checker, stmt: &Stmt) {
+    let Some(appends) = match_consecutive_appends(checker, stmt) else {
+        return;
+    };
+
+    // No need to proceed if we have less than 1 append to work with.
+    if appends.len() <= 1 {
+        return;
+    }
+
+    let groups = group_appends(&appends);
+
+    // group borrows from checker, so we can't directly push into checker.diagnostics
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+    diagnostics.reserve_exact(groups.len());
+
+    for group in groups {
+        let replacement = make_suggestion(&group, checker.generator());
+        let start = group.appends.first().unwrap().stmt.start();
+        let end = group.appends.last().unwrap().stmt.end();
+
+        let mut diagnostic = Diagnostic::new(
+            RepeatedAppend {
+                name: group.name().to_string(),
+                replacement: replacement.to_string(),
+            },
+            TextRange::new(start, end),
+        );
+
+        // We only suggest fix when all appends in a group are clumped together.
+        if checker.patch(diagnostic.kind.rule()) && group.is_consecutive {
+            diagnostic.set_fix(Fix::suggested(Edit::replacement(replacement, start, end)));
+        }
+
+        diagnostics.push(diagnostic);
+    }
+    checker.diagnostics.extend(diagnostics);
+}
+
+#[derive(Debug, Clone)]
+struct Append<'a> {
+    receiver: &'a ExprName,
+    binding: &'a Binding<'a>,
+    argument: &'a Expr,
+    stmt: &'a Stmt,
+}
+
+#[derive(Debug)]
+struct AppendGroup<'a> {
+    appends: Vec<Append<'a>>,
+    is_consecutive: bool,
+}
+
+impl<'a> AppendGroup<'a> {
+    fn name(&self) -> &'a str {
+        &self.appends.first().unwrap().receiver.id
+    }
+}
+
+fn match_consecutive_appends<'a>(checker: &'a Checker, stmt: &'a Stmt) -> Option<Vec<Append<'a>>> {
+    // In order to match consecutive statements, we need to go to the tree ancestor of the
+    // given statement, find its position there, and match all 'appends' from there.
+    let siblings: &'a [Stmt] = if checker.semantic().at_top_level() {
+        // If the statement is at the top level, we should go to the parent module.
+        // Module is available in the definitions list.
+        let Definition::Module(ref module_def) =
+            checker.semantic().definitions[DefinitionId::module()]
+        else {
+            return None;
+        };
+        module_def.python_ast
+    } else {
+        // Otherwise, go to the parent, and take its body as a sequence of siblings.
+        let parent = checker.semantic().current_statement_parent()?;
+        traversal::suite(stmt, parent)?
+    };
+
+    let stmt_index = siblings.iter().position(|x| x == stmt)?;
```

Comment:
Maybe we can add a method to `traversal` to support iterating over the siblings? We already have `next_sibling`.

Reply:
Do you want me to extend traversal to include the logic about the global scope statements?

Reply:
`traversal::suite` already returns siblings, but doesn't handle the situation with the top level statements. Right now traversal is part of AST, but in order to resolve this issue, it needs access to semantics. I can move it to a different crate and make that change. 

It can probably be a separate PR, though. What do you think it should be?